File: di_java/src/test/java/com/kmmania/api/DAOTest.java
package com.kmmania.api;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.*;

public class DAOTest {
    @InjectMocks
    private DAO daoMock;

    @BeforeEach
    void setUp() {
        daoMock = mock(DAO.class);
    }

    @Test
    @DisplayName("getData() method should return 10")
    void getDataShouldReturn10() {
        // Given
        when(daoMock.getData()).thenReturn(10.0);

        // When
        double result = daoMock.getData();

        // Then
        verify(daoMock, times(1)).getData();
        assertThat(result).isEqualTo(10.0);
    }

    @Test
    @DisplayName("Direct instanciation should throw exception")
    void directInstantiationShouldThrowException() {
        // Assert that attempting to directly instantiate DAO throws an exception
        assertThrows(UnsupportedOperationException.class, () -> {
            DAO dao = () -> {
                throw new UnsupportedOperationException(
                        "DAO is an interface and cannot be instantiated directly."
                );
            };
            dao.getData(); // This should throw UnsupportedOperationException
        });
    }
}

File: di_kotlin/src/test/kotlin/api/DAOTest.kt
package api

import com.kmmania.api.DAO
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.InjectMocks
import org.mockito.Mockito.*

class DAOTest {
    @InjectMocks
    private lateinit var daoMock: DAO

    @BeforeEach
    fun setUp() {
        daoMock = mock(DAO::class.java)
    }

    @Test
    fun `getData() method should return 10`() {
        // Given
        `when`(daoMock.getData()).thenReturn(10.0)

        // When
        val result = daoMock.getData()

        // Then
        verify(daoMock, times(1)).getData()
        assertThat(result).isEqualTo(10.0)
    }

    @Test
    fun `Direct instanciation should throw exception`() {
        // Assert that attempting to directly instantiate DAO throws an exception
        assertThrows(UnsupportedOperationException::class.java) {
            doThrow(
                UnsupportedOperationException(
                    "DAO is an interface and cannot be instantiated directly."
                )
            ).`when`(daoMock).getData()

            daoMock.getData() // This should throw UnsupportedOperationException
        }
    }
}